// @zappar/video-recorder/lib/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getResolvedOptions = void 0;
function getResolvedOptions(opts) {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    return {
        autoFrameUpdate: (_a = opts.autoFrameUpdate) !== null && _a !== void 0 ? _a : true,
        maxFrameRate: (_b = opts.maxFrameRate) !== null && _b !== void 0 ? _b : 30,
        speed: (_c = opts.speed) !== null && _c !== void 0 ? _c : 10,
        quality: (_d = opts.quality) !== null && _d !== void 0 ? _d : 25,
        audio: (_e = opts.audio) !== null && _e !== void 0 ? _e : true,
        halfSample: (_f = opts.halfSample) !== null && _f !== void 0 ? _f : false,
        targetBitrateKBPS: (_g = opts.targetBitrateKBPS) !== null && _g !== void 0 ? _g : 0,
        audioSampleRate: (_h = opts.audioSampleRate) !== null && _h !== void 0 ? _h : 32000,
        audioMediaStream: opts.audioMediaStream,
        audioSourceNode: opts.audioSourceNode,
        audioContext: opts.audioContext
    };
}
exports.getResolvedOptions = getResolvedOptions;
